### PR TITLE
Introduce real product type field

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Up until now Product had no persisted type, its type was deduced based on absence/presence of some associations The idea is now to save this type in DB and make all the rules more consistent (prevent actions not allowed for certain types) It also enables us to have a more stable dynamic form which content adapts based on this product type<br><br>Finally in the product page we force the merchant to submit the form straight away and we display some warning to explain that chaging the type will erase some data (packs associations, combinations, virtual file)
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | CI tests green
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23702)

This is what the modal confirmation looks like (for now) when you change the type of the product:

Capture d’écran 2021-03-23 à 23 11 13

### CODE REVIEW

**Comment by kpodemski** [comment]
I'm wondering if we should think about module developers here, "Subscription product", "Preorder product" and few other scenarios where I recall special modules for it 🤔

File: src/Adapter/Product/Combination/Repository/CombinationRepository.php [hunk 1]
```diff
@@ -143,14 +144,72 @@ public function partialUpdate(Combination $combination, array $updatableProperti
...
+    {
+        $combinationIds = $this->getCombinationIdsByProductId($productId);
+
+        $this->bulkDelete($combinationIds);
+    }
+
+    /**
+     * @param array $combinationIds
+     */
+    public function bulkDelete(array $combinationIds): void
```
Comment (jolelievre) [inline src/Adapter/Product/Combination/Repository/CombinationRepository.php:172]: Initially I performed a bulk delete with SQL only, but we need to use the ObjectModel because some hooks are triggered And without them some modules (and their tables) are not correctly cleaned (like index table from faceted search) which causes bugs afterwards

Comment (jolelievre) [reply src/Adapter/Product/Combination/Repository/CombinationRepository.php:172]: The limitation here will be when the product has hundreds or thousands of combinations

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    return {};
+  }
+
+  /**
+   * @private
+   */
+  confirmTypeSubmit() {
+    let confirmMessage = this.$typeSelector.data('confirm-message');
+    let confirmWarning;
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:54]:
```suggestion
    let confirmWarning = '';
```

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        break;
+      case ProductMap.productType.PACK:
+        confirmWarning = this.$typeSelector.data('pack-warning');
+        break;
+      case ProductMap.productType.VIRTUAL:
+        confirmWarning = this.$typeSelector.data('virtual-warning');
+        break;
+      case ProductMap.productType.STANDARD:
+      default:
+        confirmWarning = '';
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:68]:
```suggestion
```

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        break;
+    }
+
+    if (confirmWarning) {
+      confirmWarning = `<div class="alert alert-warning">${confirmWarning}</div>`;
+    }
+    confirmMessage = `<div class="alert alert-info">${confirmMessage}</div>`;
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:75]:
```suggestion
    if (confirmMessage) {
      confirmMessage = `<div class="alert alert-info">${confirmMessage}</div>`;
    }
```

File: classes/Product.php [hunk 5]
```diff
@@ -471,6 +477,12 @@ class ProductCore extends ObjectModel
                 'validate' => 'isGenericName',
                 'size' => 255,
             ],
+            'product_type' => [
+                'type' => self::TYPE_STRING,
+                'validate' => 'IsGenericName',
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
                'validate' => 'isGenericName',
```

File: install-dev/upgrade/php/ps_1780_update_product_type.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+function ps_1780_update_product_type()
```
Comment (Progi1984) [inline install-dev/upgrade/php/ps_1780_update_product_type.php]: Why put in a php file like it's just sql requests ?

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePackFeatureContext.php [hunk 7]
```diff
@@ -84,6 +85,20 @@ public function removeAllProductsFromPack(string $packReference): void
...
+     *
+     * @param string $packReference
+     */
+    public function assertPackEmpty(string $packReference): void
+    {
+        $packId = $this->getSharedStorage()->get($packReference);
+
+        $packedProducts = $this->getQueryBus()->handle(new GetPackedProducts($packId));
+        Assert::assertEmpty($packedProducts);
+        Assert::assertFalse(Pack::isPack($packId));
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePackFeatureContext.php:99]: How a pack cannot be a pack ?

File: src/Core/Domain/Product/CommandHandler/UpdateProductTypeHandlerInterface.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
```
Comment (Progi1984) [inline src/Core/Domain/Product/CommandHandler/UpdateProductTypeHandlerInterface.php]:
```suggestion
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js (hunk 3 above)
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:68]: Actually I prefer the way I used, because I remember the lint displaying some warning about missing default state sometimes Besides this way the use case is clear, so is the default value, and it's not less performant

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js (hunk 4 above)
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:75]: Nope confirmMessage is always there, the if is useless

File: install-dev/upgrade/php/ps_1780_update_product_type.php (hunk 6 above)
Comment (jolelievre) [reply install-dev/upgrade/php/ps_1780_update_product_type.php]: Oh you're right, at first I created the PHP file but left it empty because I was expecting some logic to handle this And I just though of this trick yesterday so I added it here But indeed it can all be in SQL which is better and simpler

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePackFeatureContext.php (hunk 7 above)
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePackFeatureContext.php:99]: Because `Pack::isPack` is a method that returns the presence of pack associations Up until now we only relied on the presence/absence of this associations to define the type of the product

That's the whole point of the PR actually (explained in the description) we do not rely only on associations any more but on a field that represents (and persist) the product type

TLDR; when is a pack not a pack? When its product_type field is equal to pack but the merchant has not defined associations with other products yet

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePackFeatureContext.php:99]: Or when you just removed all its products, then no association but the `product_type` is still pack

**Comment by jolelievre** [comment]
Thanks @Progi1984 you feedback have been addressed 😉

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js (hunk 4 above)
Comment (Progi1984) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:75]: Yeap but it can be empty as we fetch data...

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:75]: Well in this case the content will be empty, but it doesn't prevent from wrapping it Because it's not supposed to happen this component is only linked to one form field it's not reusable And the data attribute is correctly set in in `BasicInformationType` right now

For a reusable dynamic component it would worth adding extra checking but in this case I don't think it's valuable If we did this for all our components and pages we would have many ifs everywhere

Last note: even if it's empty it won't throw any error anyway it will be just less pretty

**Review by Progi1984 — DISMISSED** [review]

File: admin-dev/themes/new-theme/js/pages/product/edit/index.js [hunk 9]
```diff
@@ -49,9 +50,9 @@ $(() => {
   const productId = parseInt($productForm.data('productId'), 10);
   const productType = $productForm.data('productType');
 
-  // Combinations manager must be initialised before nav handler, or it won't trigger the pagination if the tab is
-  // selected on load, but only when productId exists (edition mode)
   if (productId) {
+    // Combinations manager must be initialised before nav handler, or it won't trigger the pagination if the tab is
```
Comment (sowbiba) [inline admin-dev/themes/new-theme/js/pages/product/edit/index.js:54]:
```suggestion
    // Combinations manager must be initialized before nav handler, or it won't trigger the pagination if the tab is
```

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        confirmWarning = this.$typeSelector.data('virtual-warning');
+        break;
+      case ProductMap.productType.STANDARD:
+      default:
+        confirmWarning = '';
+        break;
+    }
+
+    if (confirmWarning) {
+      confirmWarning = `<div class="alert alert-warning">${confirmWarning}</div>`;
```
Comment (sowbiba) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:73]: $('div).addClass('alert alert-warning').html(confirmWarning)

but ... the result is the same

**Review by Progi1984 — APPROVED** [review]

**Comment by jolelievre** [comment]
> I'm wondering if we should think about module developers here, "Subscription product", "Preorder product" and few other scenarios where I recall special modules for it 🤔

@kpodemski that's interesting, so you're saying some modules may have other types of product? And the problem here is that the number of types is finite and can't be extended right?

Although this type didn't exist previously so how did the modules handle them? I would guess it was additional feature added by the module But were they added in the type selector as well? It would be interesting to analyze such modules to see how we could offer overriding and extending possibilities for this new type

Keep in mind that this page is a "beta/feature" page and will keep improving until 1.7.9 so 1.7.8 will be a good opportunity for module developers to point at what is missing and what will have to be improved. But it's definitely a point we have to remember about.

**Review by sowbiba — APPROVED** [review]

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js (hunk 10 above)
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-type-manager.js:73]: yep same difference ^^, besides the ConfirmModal expects a string not a jQuery object

File: src/PrestaShopBundle/Form/Admin/Sell/Product/BasicInformationType.php [hunk 11]
```diff
@@ -54,11 +54,18 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     $this->trans('Standard product', 'Admin.Catalog.Feature') => ProductType::TYPE_STANDARD,
                     $this->trans('Pack of products', 'Admin.Catalog.Feature') => ProductType::TYPE_PACK,
                     $this->trans('Virtual product', 'Admin.Catalog.Feature') => ProductType::TYPE_VIRTUAL,
-                    $this->trans('Product with combinations', 'Admin.Catalog.Feature') => ProductType::TYPE_COMBINATION,
+                    $this->trans('Product with combinations', 'Admin.Catalog.Feature') => ProductType::TYPE_COMBINATIONS,
                 ],
                 'choice_translation_domain' => 'Admin.Catalog.Feature',
                 'attr' => [
                     'class' => 'custom-select',
+                    'data-modal-title' => $this->trans('Changing product type', 'Admin.Catalog.Notification'),
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/BasicInformationType.php:62]: @Julievrz I'd like your feedback regarding these new wordings Seems like the bot didn't detect them for some reason

**Comment by Progi1984** [comment]
@TristanLDD : Could you validate the UX for this ?

> Capture d’écran 2021-03-23 à 23 11 13

**Comment by jolelievre** [comment]
> @TristanLDD : Could you validate the UX for this ?
> 
> > Capture d’écran 2021-03-23 à 23 11 13

If it can be done in another PR it would be great, this one is quite huge and critical But I agree the UX has (little) room for improvement 😅

**Comment by Julievrz** [comment]
@jolelievre 

Here's my suggestion:

'Are you sure you want to change the product type?' Admin.Catalog.Notification

'Changing the product type will immediately save the product and refresh the page.' Admin.Catalog.Notification

'This will delete all combinations.' 
'This will delete the list of products in this pack.' 
'This will delete the associated virtual file.' in Admin.Catalog.Notification

2 exit buttons:
'Change product type' Admin.Catalog.Notification
'Cancel' Admin.Actions
